src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs

```diff
@@ -0,0 +1,276 @@
+
+namespace Avalonia.Animation.Keyframes
+{
+    /// <summary>
+    /// Provides statefulness for an iteration of a keyframe group animation.
+    /// </summary>
+    internal class KeyFramesStateMachine<T> : IObservable<object>, IDisposable
+    {
+        object _lastInterpValue;
+        object _firstKFValue;
+        
+        private ulong _delayTotalFrameCount,
+            _durationTotalFrameCount,
+            _delayFrameCount,
+            _durationFrameCount,
+            _repeatCount,
+            _currentIteration;
+
+        private bool _isLooping,
+                     _isRepeating,
+                     _isReversed,
+                     _checkLoopAndRepeat,
+                     _gotFirstKFValue;
+        private FillMode _fillMode;
+        private PlaybackDirection _animationDirection;
+        private KeyFramesStates _currentState, _savedState;
+        private KeyFrames<T> _parent;
+        private Animation _targetAnimation;
+        private Animatable _targetControl;
+        internal bool _unsubscribe = false;
+        private IObserver<object> _targetObserver;
+
+        private enum KeyFramesStates
+        {
+            INITIALIZE,
+            DO_DELAY,
+            DO_RUN,
+            RUN_FORWARDS,
+            RUN_BACKWARDS,
+            RUN_APPLYVALUE,
+            RUN_COMPLETE,
+            PAUSE,
+            STOP,
+            DISPOSED
+        }
+
+        public void Initialize(Animation animation, Animatable control, KeyFrames<T> keyframes)
+        {
+            _parent = keyframes;
+            _targetAnimation = animation;
+            _targetControl = control;
+
+            _delayTotalFrameCount = (ulong)(animation.Delay.Ticks / Timing.FrameTick.Ticks);
+            _durationTotalFrameCount = (ulong)(animation.Duration.Ticks / Timing.FrameTick.Ticks);
+
+            switch (animation.RepeatBehavior)
+            {
+                case RepeatBehavior.Loop:
+                    _isLooping = true;
+                    _checkLoopAndRepeat = true;
+                    break;
+                case RepeatBehavior.Repeat:
+                    if (animation.RepeatCount != null)
+                    {
+                        if (animation.RepeatCount == 0)
+                        {
+                            throw new InvalidOperationException
+                                ($"RepeatCount should be greater than zero when RepeatBehavior is set to Repeat.");
+                        }
+                        _isRepeating = true;
+                        _checkLoopAndRepeat = true;
+                        _repeatCount = (ulong)animation.RepeatCount;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException
+                            ($"RepeatCount should be defined when RepeatBehavior is set to Repeat.");
+                    }
+                    break;
+            }
+
+            switch (animation.PlaybackDirection)
```

Comment:
I think this could just be:

```
_isReversed = animation.PlaybackDirection == PlaybackDirection.Reverse ||
                       animation.PlaybackDirection == PlaybackDirection.AlternateReverse;
```

Reply:
Or even make `PlaybackDirection` a `Flags` and do `_isReversed = (animation.PlaybackDirection & PlaybackDirection.Reverse) != 0;`